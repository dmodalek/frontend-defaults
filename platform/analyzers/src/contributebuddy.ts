import { directoryExists, getDependencyInstallation, DependencyInstallation } from '@namics/frontend-defaults-platform-core';
import { join } from 'path';

export type ContributebuddyAnalyzerResult = {
    contributebuddy: boolean;
    contributebuddyConfigFolder?: string,
    contributebuddyInstallation?: DependencyInstallation
};

export const contributebuddyAnalyzer = async (cwd: string): Promise<ContributebuddyAnalyzerResult> => {
    const CONTRIBUTE_BUDDY_FOLDER = join(cwd, '.contributebuddy');
    const contributebuddyConfigFolder = await directoryExists(CONTRIBUTE_BUDDY_FOLDER);
    const contributebuddyInstallation = await getDependencyInstallation(cwd, 'contribute-buddy');

    if (contributebuddyConfigFolder || contributebuddyInstallation.declared) {
        return {
            contributebuddy: true,
            contributebuddyConfigFolder: CONTRIBUTE_BUDDY_FOLDER,
            contributebuddyInstallation

        }
    }

    return {
        contributebuddy: false,
    };
}