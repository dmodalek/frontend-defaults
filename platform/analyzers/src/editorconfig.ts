import { Analyzer, fileExists } from '@namics/frontend-defaults-platform-core';
import { join } from 'path';

export type EditorconfigAnalyzerResult = {
	editorConfig: boolean;
};

export class EditorconfigAnalyzer extends Analyzer<EditorconfigAnalyzerResult> {
	async analyze(): Promise<EditorconfigAnalyzerResult> {
		const eidtorConfigExists = await fileExists(join(this.context, '.editorconfig'));

		return {
			editorConfig: eidtorConfigExists,
		};
	}
}
