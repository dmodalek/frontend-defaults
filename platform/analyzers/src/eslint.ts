import { DependencyInstallation, fileExists, getDependencyInstallation } from '@namics/frontend-defaults-platform-core';
import { join } from 'path';

export type ESLintAnalyzerResult = {
	eslint: boolean;
	eslintIgnore?: boolean;
	eslintInstallation?: DependencyInstallation
};

export const eslintAnalyzer = async (cwd: string): Promise<ESLintAnalyzerResult> => {
	const doesESLintExist = await fileExists(join(cwd, '.eslintrc.js'));

	if (doesESLintExist) {
		const doesESLintIgnoreExist = await fileExists(join(cwd, '.eslintignore'));
		const eslintInstallation = await getDependencyInstallation(cwd, 'eslint');

		return {
			eslint: doesESLintExist,
			eslintIgnore: doesESLintIgnoreExist,
			eslintInstallation
		};
	}

	return {
		eslint: doesESLintExist,
	};
}