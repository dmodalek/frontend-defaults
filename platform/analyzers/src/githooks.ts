import {
	DependencyInstallation,
	fileExists,
	getDependencyInstallation,
	getPackageJSON,
	IPackage,
} from '@namics/frontend-defaults-platform-core';
import { join } from 'path';

export type GitHooksLocation = 'rc' | 'package';

export type GitHooksAnalyzerResult = {
	gitHooks: boolean;
	gitHooksInstallation?: DependencyInstallation;
	gitHooksLocation?: GitHooksLocation;
};

type PackageWithHuskyConfig = IPackage & {
	husky: any;
};

export const gitHooksAnalyzer = async (cwd: string): Promise<GitHooksAnalyzerResult> => {
	const doesGitHooksConfigExist = await fileExists(join(cwd, '.huskyrc'));
	const gitHooksInstallation = await getDependencyInstallation(cwd, 'husky');
	const pkg: PackageWithHuskyConfig = (await getPackageJSON(cwd)) as any;

	if (doesGitHooksConfigExist || pkg.husky) {
		return {
			gitHooks: doesGitHooksConfigExist,
			gitHooksLocation: pkg.husky ? 'package' : 'rc',
			gitHooksInstallation,
		};
	}

	return {
		gitHooks: false,
	};
};
