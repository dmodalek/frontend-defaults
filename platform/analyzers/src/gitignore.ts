import { fileExists } from '@namics/frontend-defaults-platform-core';
import { join } from 'path';

export type GitIgnoreAnalyzerResult = {
	gitignore: boolean;
};

export const gitIgnoreAnalyzer = async (cwd: string): Promise<GitIgnoreAnalyzerResult> => {
	const doesGitignoreExist = await fileExists(join(cwd, '.gitignore'));

	return {
		gitignore: doesGitignoreExist,
	};
}