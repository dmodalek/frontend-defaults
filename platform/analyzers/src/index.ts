// export all related analyzer stuff for contributebuddy
export { contributebuddyAnalyzer, ContributebuddyAnalyzerResult } from './contributebuddy';

// export all related analyzer stuff for editorconfig
export { editorConfigAnalyzer, EditorconfigAnalyzerResult } from './editorconfig';

// export all related analyzer stuff for eslint
export { eslintAnalyzer, ESLintAnalyzerResult } from './eslint';

// export all related analyzer stuff for githooks
export { gitHooksAnalyzer, GitHooksAnalyzerResult } from './githooks';

// export all related analyzer stuff for gitignore
export { gitIgnoreAnalyzer, GitIgnoreAnalyzerResult } from './gitignore';

// export all related analyzer stuff for license
export { licenseAnalyzer, LicenseAnalyzerResult } from './license';

// export all related analyzer stuff for nitro
export { nitroAnalyzer, NitroAnalyzerResult } from './nitro';

// export all related analyzer stuff for node
export { nodeAnalyzer, NodeAnalyzerResult } from './node';

// export all related analyzer stuff for npmrc
export { npmrcAnalyzer, NPMRCAnalyzerResult } from './npmrc';

// export all related analyzer stuff for prettier
export { prettierAnalyzer, PrettierAnalyzerResult } from './prettier';

// export all related analyzer stuff for readme
export { readmeAnalyzer, ReadmeAnalyzerResult } from './readme';

// export all related analyzer stuff for repository
export { repositoryAnalyzer, RepositoryAnalyzerResult } from './repository';

// export all related analyzer stuff for stylelint
export { stylelintAnalyzer, StylelintAnalyzerResult } from './stylelint';

// export all related analyzer stuff for tslint
export { tslintAnalyzer, TSLintAnalyzerResult } from './tslint';

// export all related analyzer stuff for typescript
export { typescriptAnalyzer, TypeScriptAnalyzerResult } from './typescript';

// export all related analyzer stuff for webpack
export { webpackAnalyzer, WebpackAnalyzerResult } from './webpack';
