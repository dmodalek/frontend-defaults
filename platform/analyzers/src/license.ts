import { Analyzer, fileExists } from '@namics/frontend-defaults-platform-core';
import { join } from 'path';

export type LicenseAnalyzerResult = {
	license: boolean;
};

export class LicenseAnalyzer extends Analyzer<LicenseAnalyzerResult> {
	async analyze(): Promise<LicenseAnalyzerResult> {
		const licenseExists = await fileExists(join(this.context, 'LICENSE'));

		return {
			license: licenseExists,
		};
	}
}
