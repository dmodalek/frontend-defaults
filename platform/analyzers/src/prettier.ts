import {
	Analyzer,
	DependencyInstallation,
	fileExists,
	getDependencyInstallation,
	getPackageJSON,
	getFileContents,
	getJSON,
} from '@namics/frontend-defaults-platform-core';
import { join } from 'path';

type PrettierConfigLocation = 'package' | 'config';

type PrettierContent = {
	trailingComma: string;
	tabWidth: number;
	semi: boolean;
	singleQuote: boolean;
	jsxSingleQuote: boolean;
	bracketSpacing: boolean;
	jsxBracketSameLine: boolean;
	arrowParens: string;
	rangeStart: number;
	rangeEnd: number;
	requirePragma: boolean;
	insertPragma: boolean;
	proseWrap: string;
	htmlWhitespaceSensitivity: string;
};

export type PrettierAnalyzerResult = {
	prettier: boolean;
	prettierIgnore?: boolean;
	prettierSettings?: Partial<PrettierContent>;
	prettierInstallation?: DependencyInstallation;
	prettierConfigLocation?: PrettierConfigLocation;
};

export const prettierAnalyzer = async (cwd: string): Promise<PrettierAnalyzerResult> => {
	const PRETTIER_CONFIG_PATH = join(cwd, '.prettierrc.js');
	const PRETTIER_IGNORE_PATH = join(cwd, '.prettierignore');

	const prettierConfigFile = await fileExists(PRETTIER_CONFIG_PATH);
	const prettierIgnore = await fileExists(PRETTIER_IGNORE_PATH);
	const packageConfig = await getPackageJSON(cwd);
	const inlinePrettierConfig = typeof (packageConfig as any).prettier === 'object';

	// we'll prefer config files for prettier instead of inline package configuration
	// so we'll check first if a certain config file exists to use
	if (prettierConfigFile) {
		const prettierInstallation = await getDependencyInstallation(cwd, 'prettier');
		const prettierSettings = require(PRETTIER_CONFIG_PATH);

		return {
			prettier: true,
			prettierIgnore,
			prettierSettings,
			prettierInstallation,
			prettierConfigLocation: 'config',
		};
	}

	if (inlinePrettierConfig) {
		const prettierInstallation = await getDependencyInstallation(cwd, 'prettier');

		return {
			prettier: true,
			prettierIgnore,
			prettierInstallation,
			prettierSettings: (packageConfig as any).prettier,
			prettierConfigLocation: 'package',
		};
	}

	return {
		prettier: false,
	};
};
