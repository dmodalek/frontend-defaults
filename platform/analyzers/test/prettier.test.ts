import { getFixtureContext } from './utils';
import { prettierAnalyzer } from '../src/prettier';

const FIXTURE_NONE = getFixtureContext('default-project');
const FIXTURE_ALL = getFixtureContext('prettier-project');
const FIXTURE_NOT_INSTALLED = getFixtureContext('prettier-project-not-installed');
const FIXTURE_NO_IGNORE = getFixtureContext('prettier-project-no-ignore');
const FIXTURE_PACKAGE_CONFIG = getFixtureContext('prettier-project-package-config');

describe('Analyzers', () => {
    describe('PrettierAnalyzer', () => {
        it('should not crash', async () => {
            expect(async () => await prettierAnalyzer(FIXTURE_ALL)).not.toThrow();
            expect(async () => await prettierAnalyzer(FIXTURE_NOT_INSTALLED)).not.toThrow();
            expect(async () => await prettierAnalyzer(FIXTURE_NO_IGNORE)).not.toThrow();
            expect(async () => await prettierAnalyzer(FIXTURE_PACKAGE_CONFIG)).not.toThrow();
        });

        it('should find no prettier if nothing related is installed', async () => {
            const analytics = await prettierAnalyzer(FIXTURE_NONE);

            expect(analytics.prettier).toEqual(false);
            expect(analytics).toMatchSnapshot();
        });

        it('should analyze a project with installation correctly', async () => {
            const analytics = await prettierAnalyzer(FIXTURE_ALL);

            expect(analytics.prettier).toEqual(true);
            expect(analytics.prettierIgnore).toEqual(true);
            expect(analytics.prettierConfigLocation).toEqual('config');
            expect(analytics.prettierInstallation!.declared).toEqual('1.1.0');
            expect(analytics.prettierInstallation!.installed).toEqual('1.16.4');
            expect(analytics).toMatchSnapshot({
                prettierInstallation: {
                    latest: expect.any(String)
                }
            });
        });

        it('should analyze a project with config inside package.json correctly', async () => {
            const analytics = await prettierAnalyzer(FIXTURE_PACKAGE_CONFIG);

            expect(analytics.prettier).toEqual(true);
            expect(analytics.prettierIgnore).toEqual(true);
            expect(analytics.prettierConfigLocation).toEqual('package');
            expect(analytics.prettierInstallation!.declared).toEqual('1.1.0');
            expect(analytics.prettierInstallation!.installed).toEqual('1.16.4');
            expect(analytics).toMatchSnapshot({
                prettierInstallation: {
                    latest: expect.any(String)
                }
            });
        });

        it('should analyze a project without installation correctly', async () => {
            const analytics = await prettierAnalyzer(FIXTURE_NOT_INSTALLED);

            expect(analytics.prettier).toEqual(true);
            expect(analytics.prettierIgnore).toEqual(true);
            expect(analytics.prettierConfigLocation).toEqual('config');
            expect(analytics.prettierInstallation!.declared).toEqual('1.1.0');
            expect(analytics.prettierInstallation!.installed).toBeNull();
            expect(analytics).toMatchSnapshot({
                prettierInstallation: {
                    latest: expect.any(String)
                }
            });
        });

        it('should analyze a project without ignore & installation correctly', async () => {
            const analytics = await prettierAnalyzer(FIXTURE_NO_IGNORE);

            expect(analytics.prettier).toEqual(true);
            expect(analytics.prettierIgnore).toEqual(false);
            expect(analytics.prettierConfigLocation).toEqual('config');
            expect(analytics.prettierInstallation!.installed).toEqual('1.16.4');
            expect(analytics).toMatchSnapshot({
                prettierInstallation: {
                    latest: expect.any(String)
                }
            });
        });
    });
});
