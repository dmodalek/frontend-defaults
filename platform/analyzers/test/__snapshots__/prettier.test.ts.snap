// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Analyzers PrettierAnalyzer should analyze a project with config inside package.json correctly 1`] = `
Object {
  "prettier": true,
  "prettierConfigLocation": "package",
  "prettierIgnore": true,
  "prettierInstallation": Object {
    "declared": "1.1.0",
    "installed": "1.16.4",
    "latest": Any<String>,
    "location": "devDependencies",
  },
  "prettierSettings": Object {
    "printWidth": 120,
    "useTabs": true,
  },
}
`;

exports[`Analyzers PrettierAnalyzer should analyze a project with installation correctly 1`] = `
Object {
  "prettier": true,
  "prettierConfigLocation": "config",
  "prettierIgnore": true,
  "prettierInstallation": Object {
    "declared": "1.1.0",
    "installed": "1.16.4",
    "latest": Any<String>,
    "location": "devDependencies",
  },
  "prettierSettings": Object {},
}
`;

exports[`Analyzers PrettierAnalyzer should analyze a project without ignore & installation correctly 1`] = `
Object {
  "prettier": true,
  "prettierConfigLocation": "config",
  "prettierIgnore": false,
  "prettierInstallation": Object {
    "declared": "1.1.0",
    "installed": "1.16.4",
    "latest": Any<String>,
    "location": "devDependencies",
  },
  "prettierSettings": Object {},
}
`;

exports[`Analyzers PrettierAnalyzer should analyze a project without installation correctly 1`] = `
Object {
  "prettier": true,
  "prettierConfigLocation": "config",
  "prettierIgnore": true,
  "prettierInstallation": Object {
    "declared": "1.1.0",
    "installed": null,
    "latest": Any<String>,
    "location": "devDependencies",
  },
  "prettierSettings": Object {},
}
`;

exports[`Analyzers PrettierAnalyzer should find no prettier if nothing related is installed 1`] = `
Object {
  "prettier": false,
}
`;
